app/controllers/api/providers_controller.rb

```diff
@@ -230,5 +271,22 @@ def validate_provider_class
       end
       @collection_klasses[:providers] = Provider
     end
+
+    # Checks if the change password request data is valid
+    #
+    # @raise [BadRequestError] if some required data is missing
+    #                          if new_password and confirm_password are differents
+    #
+    # @return [Boolean] true if all request data is  valid
+    def validate_change_password_request_data(data)
+      if data["current_password"].blank? ||
```

Comment:
This feels unfortunate as the API now has to have some domain  knowledge about what constitutes a valid password change. IMO it would be better to pass this data to the `change_password_queue` method and allow the task to reveal what went wrong. Alternatively, you could add a service to the domain model that validates password request changes, but it's important that it lives there and not in the API layer.

Some further thoughts: is it possible for the the request to fail if it passes validation? What are the cases in which it could fail? Ideally we'd have test cases illustrating all these paths.

Reply:
Well, this validation is more about a data validation, not a domain. Here it validates if all data is provide by client and if it is well formatted (in this case, it require that the field `new_password` and `confirm_password` are equal, applying a "pattern" in change password routine, forcing the client to be sure what he is doing). The point is that this validation is more about the client/API communication and not about a domain rule, so I think its fair to have this code living here. Additionally, if you take a look on `change_password_queue` I just need to pass 2 params (`current_password` and `new_password`).

Yes, it is possible to fail if the data validation passes. To change password the data must be validate in the provide against its password policies, in failure case, the task will reveal the error.

I don't think that we can write specs for it, once it depends on provider's password rules.

Reply:
> (in this case, it require that the field new_password and confirm_password are equal, applying a "pattern" in change password routine, forcing the client to be sure what he is doing)

I disagree that this is not a domain concern - it's just one that hasn't been modeled. If some other part of the system decided to implement password changes, they'd need to duplicate this knowledge of what things are required, and how certain fields are supposed to match, etc...

There could be a number of ways to do this. One suggestion might be to wrap the password change in a service that lives outside of the API, that takes these fields and does the validation on them. Another could be that the service just does the validation, and the API makes the decision of whether to try to update the password based on the success of that. Either way, the API doesn't *know* what's required to make a password update, it's just going back and forth between the user and the domain model and responding appropriately.